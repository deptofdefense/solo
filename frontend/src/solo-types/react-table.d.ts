import {
  UseExpandedHooks,
  UseExpandedInstanceProps,
  UseExpandedOptions,
  UseExpandedRowProps,
  UseExpandedState,
  UseFiltersColumnOptions,
  UseFiltersColumnProps,
  UseFiltersInstanceProps,
  UseFiltersOptions,
  UseFiltersState,
  UseGlobalFiltersInstanceProps,
  UseGlobalFiltersOptions,
  UseGlobalFiltersState,
  UsePaginationInstanceProps,
  UsePaginationOptions,
  UsePaginationState,
  UseRowStateCellProps,
  UseRowStateInstanceProps,
  UseRowStateOptions,
  UseRowStateRowProps,
  UseRowStateState,
  UseSortByColumnOptions,
  UseSortByColumnProps,
  UseSortByHooks,
  UseSortByInstanceProps,
  UseSortByOptions,
  UseSortByState
} from "react-table";

declare module "react-table" {
  export interface TableOptions<D extends object>
    extends UseExpandedOptions<D>,
      UseFiltersOptions<D>,
      UseGlobalFiltersOptions<D>,
      UsePaginationOptions<D>,
      UseRowStateOptions<D>,
      UseSortByOptions<D>,
      Record<string, any> {}

  export interface Hooks<D extends object = {}>
    extends UseExpandedHooks<D>,
      UseSortByHooks<D> {}

  export interface TableInstance<D extends object = {}>
    extends UseColumnOrderInstanceProps<D>,
      UseExpandedInstanceProps<D>,
      UseFiltersInstanceProps<D>,
      UseGlobalFiltersInstanceProps<D>,
      UsePaginationInstanceProps<D>,
      UseRowStateInstanceProps<D>,
      UseSortByInstanceProps<D> {}

  export interface TableState<D extends object = {}>
    extends UseColumnOrderState<D>,
      UseExpandedState<D>,
      UseFiltersState<D>,
      UseGlobalFiltersState<D>,
      UsePaginationState<D>,
      UseRowStateState<D>,
      UseSortByState<D> {}

  export interface Column<D extends object = {}>
    extends UseFiltersColumnOptions<D>,
      UseSortByColumnOptions<D> {}

  export interface ColumnInstance<D extends object = {}>
    extends UseFiltersColumnProps<D>,
      UseSortByColumnProps<D> {}

  export interface Cell<D extends object = {}>
    extends UseRowStateCellProps<D> {}

  export interface Row<D extends object = {}>
    extends UseExpandedRowProps<D>,
      UseRowStateRowProps<D> {}
}
