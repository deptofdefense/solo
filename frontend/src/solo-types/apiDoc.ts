import { ApiDocument, Status } from "solo-types";
import * as faker from "faker";

export const defaultApiDoc: ApiDocument = {
  id: 1,
  statuses: [
    {
      id: 1,
      dic: {
        id: 1,
        code: "AE1",
        desc: "in transit"
      },
      status_date: "2020-03-01T21:47:13-05:00",
      key_and_transmit_date: null,
      esd: "2020-03-20",
      projected_qty: 2,
      received_qty: null,
      document: 1
    },
    {
      id: 2,
      dic: {
        id: 2,
        code: "AS1",
        desc: "received"
      },
      status_date: "2020-03-05T21:47:56-05:00",
      key_and_transmit_date: null,
      esd: null,
      projected_qty: 2,
      received_qty: null,
      document: 1
    }
  ],
  suppadd: {
    id: 1,
    subinventorys: [
      {
        id: 1,
        locators: [
          {
            id: 1,
            code: "M1234AA",
            desc: "",
            subinventorys: 1
          },
          {
            id: 2,
            code: "M1235A",
            desc: "",
            subinventorys: 1
          },
          {
            id: 3,
            code: "M1236",
            desc: "",
            subinventorys: 1
          }
        ],
        code: "MTM_STGE",
        desc: "",
        suppadd: 1
      },
      {
        id: 2,
        locators: [],
        code: "MTM_LAY",
        desc: "",
        suppadd: 1
      },
      {
        id: 3,
        locators: [],
        code: "MTM_PEB",
        desc: "",
        suppadd: 1
      }
    ],
    code: "YMTM",
    desc: "Motor Transportation"
  },
  part: {
    id: 2,
    nsn: "5430015061999",
    nomen: "TANK,LIQUID STORAGE",
    uom: "ea",
    price: 15,
    sac: 1,
    serial_control_flag: "n",
    lot_control_flag: "n",
    recoverability_code: "z",
    shelf_life_code: 0,
    controlled_inv_item_code: "u"
  },
  service_request: {
    id: 1,
    service_request: "12345678"
  },
  addresses: [
    {
      id: 4,
      address_type: {
        id: 4,
        type: "Bill-To",
        desc: null
      },
      name: "AAC-M30300",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "bldg 24006 Montezuma",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 3,
      address_type: {
        id: 3,
        type: "Requester",
        desc: null
      },
      name: "requestor",
      ric: "5E3",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 2,
      address_type: {
        id: 2,
        type: "Ship-To",
        desc: null
      },
      name: "ship to",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 1,
      address_type: {
        id: 1,
        type: "Holder",
        desc: null
      },
      name: "AAC-M30300",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    }
  ],
  sdn: "M3030012341234"
};

const idxToStatus = (idx: number): Status => {
  const base: Omit<Status, "dic"> = {
    id: Math.floor(Math.random() * 20),
    status_date: faker.date.recent().toISOString(),
    key_and_transmit_date: null,
    esd: "2020-03-20",
    projected_qty: 20,
    received_qty: faker.random.number(20),
    document: 1
  };
  switch (idx) {
    case 0:
      return {
        ...base,
        dic: {
          id: 1,
          code: "AE1",
          desc: "in transit"
        }
      };
    case 1:
      return {
        ...base,
        dic: {
          id: 2,
          code: "AS1",
          desc: "shipped"
        }
      };
    case 2:
      return {
        ...base,
        dic: {
          id: 3,
          code: "AS2",
          desc: "in transit"
        }
      };
    case 3:
      return {
        ...base,
        dic: {
          id: 4,
          code: "D6T",
          desc: "received"
        }
      };
    default:
      return {
        ...base,
        dic: {
          id: 5,
          code: "COR",
          desc: "confirmed"
        }
      };
  }
};

const createFakeApiDocuments = (count: number = 20): ApiDocument[] => {
  return Array.from({ length: count }).map(() => ({
    ...defaultApiDoc,
    statuses: Array.from({
      length: faker.random.number({ min: 1, max: 5 })
    }).map((_, idx) => idxToStatus(idx)),
    service_request: {
      id: faker.random.number(50),
      service_request: faker.random.alphaNumeric(8).toUpperCase()
    },
    part: {
      ...defaultApiDoc.part,
      nomen: faker.commerce.productName()
    },
    suppadd: {
      ...defaultApiDoc.suppadd,
      desc: faker.random.arrayElement([
        "Motor Transportation",
        "Armory",
        "Communications",
        "Electronics Maintenence"
      ])
    },
    addresses: defaultApiDoc.addresses.map(addy => ({
      ...addy,
      name: faker.company.companyName()
    }))
  }));
};

export default createFakeApiDocuments;
