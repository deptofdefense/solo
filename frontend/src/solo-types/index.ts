import { SortingRule, Filters } from "react-table";
export {
  default as createFakeApiDocs,
  defaultApiDoc,
  defaultApiResponse
} from "./apiDoc";
export * from "./warehouseUsers";
export { default as createFakeDocs, defaultDoc } from "./doc";

export interface Part {
  id: number;
  nsn: string;
  nomen: string;
  uom: string;
  price?: number;
  sac?: number;
  serial_control_flag?: string;
  lot_control_flag?: string;
  recoverability_code?: string;
  shelf_life_code?: number;
  controlled_inv_item_code?: string;
}

export interface Dic {
  id: number;
  code: string;
  desc: string;
}

export interface Status {
  id: number;
  dic: Dic;
  status_date: string;
  key_and_transmit_date?: string | null;
  esd?: string | null;
  received_qty: number | null;
  projected_qty: number;
  document: number;
  user?: number | null;
}

export interface AddressType {
  id: number;
  type: string;
  desc?: string | null;
}

export interface Address {
  id: number;
  address_type: AddressType;
  name: string;
  ric: string;
  addy1: string;
  addy2: string;
  addy3: string;
  city: string;
  state: string;
  zip: string;
  country: string;
  document: number[];
}

export interface ServiceRequest {
  id: number;
  service_request: string;
}

export interface Locator {
  id: number;
  code: string;
  desc?: string;
  subinventorys?: number;
}

export interface Subinventory {
  id: number;
  locators: Locator[];
  code: string;
  desc?: string;
  suppadd?: number;
}

export type LocatorMap = Record<string, Locator[]>;

export interface SuppAdd {
  id: number;
  code: string;
  subinventorys: Subinventory[];
}

interface BaseDocument {
  id: number;
  sdn: string;
  part: Part;
  statuses: Status[];
}

export interface ApiDocument extends BaseDocument {
  service_request: ServiceRequest;
  addresses: Address[];
  suppadd: SuppAdd;
}

export interface Document extends BaseDocument {
  serviceRequest: ServiceRequest;
  shipper?: Address;
  receiver?: Address;
  loadingStatus: LoadingStatus;
  subinventorys: Omit<Subinventory, "locators">[];
  locatorsBySubinventory: LocatorMap;
  commodityName: string;
  mostRecentStatusIdx: number;
  enteredReceivedQty: number;
  enteredReceivedBy: string;
  enteredSubinventoryCode: string;
  enteredLocatorCode: string;
}

export interface LoadingStatus {
  loading: boolean;
  error?: boolean;
  message?: string;
}

export interface Query<T extends object> {
  sort: SortingRule<T>[];
  page: number;
  filters: Filters<T>;
}

export interface PaginatedApiResponse<T> {
  results: T;
  count: number;
  next: number;
  previous: number;
}

export interface WarehouseUser extends LoadingStatus {
  userId: number;
  username: string;
  canD6T: boolean;
  canCOR: boolean;
  aac: string;
  hasModified?: boolean;
}
