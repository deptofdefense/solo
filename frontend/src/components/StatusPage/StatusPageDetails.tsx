import React from "react";
import { Document } from "solo-types";
import StatusPageDetailData from "./StatusPageDetailData";

interface StatusPageDetailsProps {
  document: Document;
}

const StatusPageDetails: React.FC<StatusPageDetailsProps> = ({ document }) => {
  const { status, part, address } = document;

  return (
    <>
      <StatusPageDetailData address={address} part={part} status={status} />
    </>
  );
};

export default StatusPageDetails;
