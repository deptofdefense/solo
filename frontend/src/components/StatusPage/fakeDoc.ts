import { Document } from "solo-types";
import { parseISO, sub } from "date-fns";

export const defaultDoc = {
  id: 1,
  part: [
    {
      id: 2,
      nsn: "5930015931045",
      nomen: "Switch, rotary, 24V dc waterproof",
      uom: "ea",
      document: [1]
    }
  ],
  status: [
    {
      id: 1,
      dic: {
        id: 1,
        code: "AS1",
        desc: "Shipping Status",
        status: [1]
      },
      status_date: "2020-03-06T13:01:21-05:00",
      key_and_transmit_date: null,
      esd: null,
      qty: 5,
      document: 1,
      user: null
    },
    {
      id: 2,
      dic: {
        id: 1,
        code: "AS2",
        desc: "Shipping Status",
        status: [1]
      },
      status_date: "2020-03-06T13:01:21-05:00",
      key_and_transmit_date: null,
      esd: null,
      qty: 5,
      document: 1,
      user: null
    }
  ],
  address: [
    {
      id: 7,
      address_type: "1",
      name: "requestor",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 8,
      address_type: "2",
      name: "AAC-Residence Inn Arlington Pentagon City: Rm 1103",
      ric: "5E3",
      addy1: "Residence Inn Arlington Pentagon City: Rm 1103",
      addy2: "550 Army Navy Dr",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 9,
      address_type: "3",
      name: "AAC-M30300",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 10,
      address_type: "4",
      name: "AAC-M30300",
      ric: "5E3",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "bldg 24006 Montezuma",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    }
  ],
  sdn: "M0000000000000",
  service_request: 5
};

const createFakeDocuments = (count: number = 20): Document[] => {
  return Array.from({ length: count }).map(() => ({
    ...defaultDoc,
    status: defaultDoc.status.map(stat => ({
      ...stat,
      status_date: sub(parseISO(stat.status_date), {
        days: Math.floor(Math.random() * 10)
      }).toISOString()
    }))
  }));
};

export default createFakeDocuments;
