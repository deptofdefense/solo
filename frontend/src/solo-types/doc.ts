import { Document } from "solo-types";

export const defaultDoc: Document = {
  id: 1,
  statuses: [
    {
      id: 1,
      dic: {
        id: 1,
        code: "AS1",
        desc: ""
      },
      status_date: "2020-03-01T21:47:13-05:00",
      key_and_transmit_date: null,
      esd: "2020-03-20",
      projected_qty: 2,
      received_qty: null,
      document: 1
    },
    {
      id: 2,
      dic: {
        id: 2,
        code: "AB1",
        desc: ""
      },
      status_date: "2020-03-05T21:47:56-05:00",
      key_and_transmit_date: null,
      esd: null,
      projected_qty: 2,
      received_qty: null,
      document: 1
    }
  ],
  part: {
    id: 2,
    nsn: "5430015061999",
    nomen: "TANK,LIQUID STORAGE",
    uom: "ea",
    price: 15,
    sac: 1,
    serial_control_flag: "n",
    lot_control_flag: "n",
    recoverability_code: "z",
    shelf_life_code: 0,
    controlled_inv_item_code: "u"
  },
  serviceRequest: {
    id: 1,
    service_request: "12345678"
  },
  suppadd: {
    id: 5,
    code: "YMTM",
    desc: "Motor Transportation",
    subinventorys: []
  },
  shipper: {
    id: 4,
    address_type: {
      id: 4,
      type: "Bill-To",
      desc: null
    },
    name: "AAC-M30300",
    ric: "SMS",
    addy1: "addy1",
    addy2: "addy2",
    addy3: "bldg 24006 Montezuma",
    city: "Arlington",
    state: "VA",
    zip: "22202",
    country: "United States",
    document: [1]
  },
  receiver: {
    id: 3,
    address_type: {
      id: 3,
      type: "Requester",
      desc: null
    },
    name: "requestor",
    ric: "5E3",
    addy1: "addy1",
    addy2: "addy2",
    addy3: "addy3",
    city: "Arlington",
    state: "VA",
    zip: "22202",
    country: "United States",
    document: [1]
  },
  sdn: "M3030012341234",
  commodityName: "Armory",
  loadingStatus: {
    loading: false
  },
  mostRecentStatusIdx: 1,
  enteredReceivedBy: "",
  enteredReceivedQty: 2
};

const createFakeDocuments = (count: number = 20): Document[] => {
  return Array.from({ length: count }).map(() => ({
    ...defaultDoc,
    id: Math.floor(Math.random() * 100)
  }));
};

export default createFakeDocuments;
