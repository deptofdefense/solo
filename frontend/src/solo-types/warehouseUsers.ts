import { WarehouseUser, PaginatedApiResponse } from "solo-types";
import * as faker from "faker";

export const defaultWareHouseUser: WarehouseUser = {
  userId: 1,
  username: "1234123123",
  first_name: "testFirst",
  last_name: "testLast",
  aac: "testaac",
  canD6T: false,
  canCOR: false,
  loading: false
};

export const defaultUserApiResponse: PaginatedApiResponse<WarehouseUser[]> = {
  results: [defaultWareHouseUser],
  count: 10,
  next: 2,
  previous: 1
};

export const createFakeUsers = (count: number = 25): WarehouseUser[] => {
  return Array.from({ length: count }).map((_, idx) => ({
    ...defaultWareHouseUser,
    userId: idx + 1,
    username: faker.finance.account(10),
    first_name: faker.name.firstName(),
    last_name: faker.name.lastName(),
    aac: faker.random.arrayElement(["M30300", "M30305"]),
    canCOR: faker.random.boolean(),
    canD6T: faker.random.boolean()
  }));
};
