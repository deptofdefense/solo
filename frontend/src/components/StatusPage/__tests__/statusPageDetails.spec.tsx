import React from "react";
import StatusPageDetails from "../StatusPageDetails";
import StatusPageDetailsData from "../StatusPageDetailData";
import { defaultDoc } from "../fakeDoc";
import { Address, Part, Status } from "solo-types";
import { render } from "test-utils";

jest.mock("date-fns", () => ({
  parseISO: () => "testiso",
  formatDistanceToNow: () => {
    return "some amount of time";
  }
}));

describe("StatusPageDetails Component", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<StatusPageDetails document={defaultDoc} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("StatusPageDetailsData Component", () => {
  const {
    status: validStatus,
    address: validAddress,
    part: validPart
  } = defaultDoc;

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it("excludes shipper and receiver information if not available", () => {
    const { queryByText } = render(
      <StatusPageDetailsData
        address={[]}
        status={validStatus}
        part={validPart}
      />
    );
    expect(queryByText("Shipped From")).not.toBeInTheDocument();
  });

  it("includes shipper information if avaialable", () => {
    const shipper: Address = {
      ...validAddress[0],
      address_type: "2",
      name: "test shipper"
    };
    const { queryByText } = render(
      <StatusPageDetailsData
        address={[shipper]}
        status={validStatus}
        part={validPart}
      />
    );
    expect(queryByText("Shipped From")).toBeInTheDocument();
    expect(queryByText("test shipper")).toBeInTheDocument();
  });

  it("includes receiver information if avaialable", () => {
    const shipper: Address = {
      ...validAddress[0],
      address_type: "3",
      name: "test receiver"
    };
    const { queryByText } = render(
      <StatusPageDetailsData
        address={[shipper]}
        status={validStatus}
        part={validPart}
      />
    );
    expect(queryByText("Shipped To")).toBeInTheDocument();
    expect(queryByText("test receiver")).toBeInTheDocument();
  });

  it("excludes part information if not available", () => {
    const { queryByText } = render(
      <StatusPageDetailsData address={[]} status={[]} part={[]} />
    );
    expect(queryByText("NIIN")).not.toBeInTheDocument();
  });

  it("includes information of first part if available", () => {
    const part: Part = {
      ...validPart[0],
      nsn: "testpartnsn",
      uom: "testunitofmeasure"
    };
    const { queryByText } = render(
      <StatusPageDetailsData address={[]} status={[]} part={[part]} />
    );
    expect(queryByText("NIIN")).toBeInTheDocument();
    expect(queryByText("Quantity")).toBeInTheDocument();
    expect(queryByText("Unit of Measure")).toBeInTheDocument();
    expect(queryByText("testpartnsn")).toBeInTheDocument();
    expect(queryByText("testunitofmeasure")).toBeInTheDocument();
  });

  it("formats status date relative to the current time", () => {
    const status: Status = {
      ...validStatus[0]
    };
    const { queryByText } = render(
      <StatusPageDetailsData address={[]} status={[status]} part={[]} />
    );
    expect(queryByText("some amount of time ago")).toBeInTheDocument();
  });
});
