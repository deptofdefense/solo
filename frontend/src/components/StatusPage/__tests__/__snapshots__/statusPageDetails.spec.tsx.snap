// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusPageDetails Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-row flex-justify flex-align-start padding-5"
  >
    <div
      class="flex-col"
    >
      <div
        class="text-bold"
      >
        DIC
      </div>
      <div
        class="grid-row flex-nowrap flex-col flex-align-center"
      >
        <div
          class="margin-right-2"
        >
          AS1
        </div>
        <div>
          some amount of time ago
        </div>
      </div>
      <div
        class="grid-row flex-nowrap flex-col flex-align-center"
      >
        <div
          class="margin-right-2"
        >
          AS2
        </div>
        <div>
          some amount of time ago
        </div>
      </div>
    </div>
    <div
      class="flex-col"
    >
      <div
        class="text-bold"
      >
        NIIN
      </div>
      <div>
        5930015931045
      </div>
    </div>
    <div
      class="flex-col"
    >
      <div
        class="text-bold"
      >
        Quantity
      </div>
      <div>
        5
      </div>
    </div>
    <div
      class="flex-col"
    >
      <div
        class="text-bold"
      >
        Unit of Measure
      </div>
      <div>
        ea
      </div>
    </div>
    <div
      class="flex-col"
    >
      <div
        class="text-bold"
      >
        Shipped From
      </div>
      <div>
        AAC-Residence Inn Arlington Pentagon City: Rm 1103
      </div>
    </div>
    <div
      class="flex-col"
    >
      <div
        class="text-bold"
      >
        Shipped To
      </div>
      <div>
        AAC-M30300
      </div>
    </div>
  </div>
</DocumentFragment>
`;
